Maintain keen-js behaviour: Only use browser xhr POST not GET

greghuc opened the pull request:
**What does this PR do?**

WARNING: This PR builds on a previous 'bug' PR I submitted today: https://github.com/keen/keen-tracking.js/pull/26. I'm opening this PR as well, to highlight some extra behavioural changes that might be worth making before a 0.0.2 release. If you're happy with this PR's behavioural changes, then feel free to merge them in asap. Alternatively, if you do merge in the [first 'bug' PR](https://github.com/keen/keen-tracking.js/pull/26) alone, I can rebase this PR on top (on Monday).

This PR changes the browser recordEvent function behaviour when config.requestType is 'xhr':
* To (new behaviour):  _only_ use XHR POST requests, regardless of payload length
* From (current behaviour): use XHR GET if payload length < 16000 chars, else POST.

Reasoning:
* The older keen-js library [only uses XHR POST requests](https://github.com/keen/keen-js/blob/940ca049d02cff10083e5d887222846f73bd9e67/src/core/lib/addEvent.js). So this PR ensures compatible behaviour between keen-js and keen-tracking.
* It's simpler, as requestType == 'xhr' only has one type of behaviour (POST only). This means it's easier to proxy the keen.io API, as the proxy only needs to implement an addEvent POST end-point, not an extra GET end-point. I ran into this problem today when I tried to switch from keen-js to keen-tracking (I use a proxy)
* It's a bit more 'correct' in REST API terms, as POST requests are generally used for updates, not GETs.

Cheers :-)

Greg

**How should this be tested? (if appropriate)**

You got me, guvnor. I didn't write any additional unit tests. I made the behavioural changes, and all the existing tests kept passing:
* run gulp with-tests
* confirm tests pass

I also manually tested all recordEvent cases:
* requestType == 'xhr'
* requestType == 'beacon' / small payload
* requestType == 'beacon' / large payload / xhr-usable
* requestType == 'beacon' / large payload / xhr-not-usable
* requestType == 'jsonp' / large payload / xhr-usable
* requestType == 'jsonp' / large payload / xhr-not-usable

They worked fine.

These changes are mainly a re-org of existing code.

**Are there any related issues open?**

Yes. This PR builds on top of my [earlier 'bug' PR](https://github.com/keen/keen-tracking.js/pull/26)

greghuc commented:
I see the build failed on 'Keen.utils.listener should set and handle `<a>` click events set with .on("click", fn):'

Not sure if this is at all related to the code changes. All the tests run fine for me locally. I saw this test fail once, but not on future test-suite runs.

dustinlarimer commented:
@greghuc thank you so much for such a comprehensive PR! :grinning: 

Your reasoning is sound – to be honest, I totally missed this detail when shoring up backward-compatibility with keen-js.

Would it be sufficient to simply change the arguments passed on line 74?
https://github.com/keen/keen-tracking.js/blob/master/lib/record-events-browser.js#L74

```javascript
// define url separately before getRequestUrl is set
var url = this.url('/events/' + encodeURIComponent(eventCollection);

// ...
    case 'xhr':
        sendXhr.call(this, 'POST', url, extendedEventBody, cb);
        break;
// ...
```

Side note: I just realized line 74's `sendXhr` call is currently incorrect.. too many arguments :)

I also just restarted that test, so we'll see if it's a test env error.. sometimes these DOM-dependent tests timeout. Could use a little tuning there, I'm sure.

greghuc commented:
> Would it be sufficient to simply change the arguments passed on line 74?

You could do that. However, I find the new form easier to understand. Means that that 3 different requestType cases are decoupled from one another. I didn't understand the original function, with its nested if+switch. I came up with the new form once I'd traced out the different cases on a piece of paper!. To compare:
* Current: https://github.com/keen/keen-tracking.js/blob/de133bf3a44912f1e08037477f3fcfcc6b108eb1/lib/record-events-browser.js#L71
* New: https://github.com/greghuc/keen-tracking.js/blob/only-use-xhr-post/lib/record-events-browser.js#L72

greghuc commented:
I should say that I don't mind if you do the one line fix :-)
The important change is the switch to XHR-POST-only.

I think the new code reads better, but I'm a sample of one..

dustinlarimer commented:
@greghuc sounds good – your code definitely reads better :)

Any thoughts on this error?

We should also add the synchronous XHR option from keen-js:
https://github.com/keen/keen-js/blob/master/CHANGELOG.md#new-to-keen-js

dustinlarimer commented:
(opened an issue for the synchronous request: #28)

greghuc commented:
Re test error, can't help much as don't have deep knowledge of this project.

From my quick investigation, I am seeing the error locally when I run 'gulp with-tests'.
The error message there is a bit more informative:

```
Uncaught AssertionError: Fake XHR onreadystatechange handler threw exception: { readyState: 4,
  requestHeaders: 
   { Authorization: '554a723d023da6cb24e51c56a9a54555e9dcf8403d4b71ffa37e9112295622e78a10eed… === null (http://localhost:9000/build/browserified-tests.js:5072)
Error: Uncaught AssertionError: Fake XHR onreadystatechange handler threw exception: { readyState: 4,
  requestHeaders: 
   { Authorization: '554a723d023da6cb24e51c56a9a54555e9dcf8403d4b71ffa37e9112295622e78a10eed… === null (build/browserified-tests.js:5072)
    at process.on.global.onerror (vendor/mocha.js:6458:10)
```

The underlying test itself seems to be just testing that a callback is called on a a.click. Nothing strange is standing out.

It's 9pm Friday where I am, so I won't be doing much more digging over the weekend.

To make progress, it might be worth merging in the [earlier bug PR re config.writePath](https://github.com/keen/keen-tracking.js/pull/26). That diminishes the code surface area that might be causing the bug. It does seem to be passing the test suite!

greghuc commented:
OK, I fixed the build, but I think the existing tests need some more work.
Want to take a look?

Firstly, the test framework is swallowing the underlacking (useful!) stacktrace. I got access to the stacktrace by:
* Searching the project for sinon.log(err.stack), and putting console.log(err.stack); before it
* I then ran gulp from the command line, opened localhost:9000 in chrome, and looked at the console.
* Which showed:
```
AssertionError: { readyState: 4,
  requestHeaders: 
   { Authorization: '554a723d023da6cb24e51c56a9a54555e9dcf8403d4b71ffa37e9112295622e78a10eed… === null
    at Function.proclaim.strictEqual (http://localhost:9000/build/browserified-tests.js:3305:13)
    at Function.proclaim.isNull (http://localhost:9000/build/browserified-tests.js:3425:18)
    at null.<anonymous> (http://localhost:9000/build/browserified-tests.js:10389:18)
    at FakeXMLHttpRequest.xhr.onreadystatechange (http://localhost:9000/build/browserified-tests.js:913:14)
    at FakeXMLHttpRequest.readyStateChange (http://localhost:9000/build/browserified-tests.js:8464:30)
    at FakeXMLHttpRequest.setResponseBody (http://localhost:9000/build/browserified-tests.js:8630:22)
    at FakeXMLHttpRequest.respond (http://localhost:9000/build/browserified-tests.js:8637:22)
    at Object.processRequest (http://localhost:9000/build/browserified-tests.js:7644:33)
    at Object.respond (http://localhost:9000/build/browserified-tests.js:7620:26)
    at Context.<anonymous> (http://localhost:9000/build/browserified-tests.js:10394:21)
```

This led to a bunch of tests in record-events-browser-spec.js. I fixed two issues:
* Removed '/events/' from ```this.postUrl = this.client.url('/events/' + config.collection);```
* Changed the test causing the stacktrace to use a POST, rather than a GET

I'm not familiar with these tests, but they need a review to check that:
* The client.url function is properly used. The suppled path parameter shouldn't have contain '/events', as that's been subsumed into the writePath() function ('/events' is the suffix of writePath)
* We've switched from using GET XHRs to POST XHRs

Hope this helps!

Greg

greghuc commented:
OK - my final comment for a Friday night!

The tests in record-events-browser-spec.js don't seem to match the reality of the code.

Specifically, the '.recordEvents' section has tests for the success/error cases of 'should send a GET request to the API using XHR'. But the code in record-events-browser.js says otherwise: the recordEvents function calls ```sendXhr.call(this, 'POST', url, extendedEventsHash, cb);```, and this wasn't something changed in this PR..

Could be the standard problem of mocks not matching reality?

dustinlarimer commented:
@greghuc this is awesome, thank you so much for digging into this! I'll review, clean stuff up, and prep a v0.0.2 release over the weekend. Sound ok?

greghuc commented:
Great, and good luck!

> On 17 Jul 2015, at 23:51, Dustin Larimer <notifications@github.com> wrote:
> 
> @greghuc <https://github.com/greghuc> this is awesome, thank you so much for digging into this! I'll review, clean stuff up, and prep a v0.0.2 release over the weekend. Sound ok?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/keen/keen-tracking.js/pull/27#issuecomment-122439249>.
>

greghuc commented:
@dustinlarimer thanks for merging these PRs :-)

dustinlarimer commented:
@greghuc no problem, thanks for the help! Should have this released shortly..

dustinlarimer commented:
@greghuc released: https://github.com/keen/keen-tracking.js/releases/tag/v0.0.2
